minikube tunnel

balopat opened the pull request:
An implementation of `minikube tunnel` based on https://github.com/kubernetes/minikube/pull/2413.

balopat commented:
cla signed.

balopat commented:
cc @dlorenc

aaron-prindle commented on cmd/minikube/cmd/tunnel.go:
```diff
@@ -0,0 +1,42 @@
+		}
+	},
+}
+
```
nit: Might make sense to add a verbose logging flag option for tunnel

aaron-prindle commented on pkg/minikube/tunnel/route_windows.go:
```diff
@@ -0,0 +1,158 @@
+	return false, nil
+}
+
+func (r *OSRouter) Cleanup() error {
```
if minikube tunnel crashes or exits unexepectedly, is there a way for users to manually run this operation, e.g. minikube tunnel --cleanup

balopat replied:
not yet, I do like the idea. Would this belong in this PR or can be done in a separate PR?

balopat replied:
I added this feature

dlorenc commented on docs/contributors/mocking.md:
```diff
@@ -0,0 +1,8 @@
+
+To mock interfaces in unit tests use [testify/mock](https://github.com/stretchr/testify) based mocks. 
+The mocks can be generated using mockery. 
+We have a wrapper script to update or create a new mock: `hack/gen_mocks.sh`.
```
Is the idea that we don't edit the mocks and they're always generated with this?

balopat replied:
yes!

balopat replied:
removed mocking.

dlorenc commented on mocks/github.com/docker/machine/libmachine_mocks/API.go:
```diff
@@ -0,0 +1,187 @@
+*/
+
+// Code generated by mockery v1.0.0. DO NOT EDIT.
+package libmachine_mocks
```
nit: underscores are frowned upon in Go package names

balopat replied:
alright, I'll change that, thank you

dlorenc commented on pkg/minikube/tunnel/loadbalancer_patcher.go:
```diff
@@ -0,0 +1,158 @@
+	var managedServices []string
+
+	for _, svc := range serviceList.Items {
+		if svc.Spec.Type == "LoadBalancer" {
```
nit: we should add some debugging in here

balopat replied:
added

dlorenc commented on pkg/minikube/tunnel/loadbalancer_patcher.go:
```diff
@@ -0,0 +1,158 @@
+		result, err = l.requestSender.Do(request)
+		logrus.Infof("Patched %s with IP %s", svc.Name, clusterIP)
+		return
+	} else {
```
nit: since you return in the if, drop the else here.

dlorenc commented on pkg/minikube/tunnel/loadbalancer_patcher.go:
```diff
@@ -0,0 +1,158 @@
+func (l *DefaultPatcher) cleanupService(restClient rest.Interface, svc core_v1.Service) (result []byte, err error) {
+	ingresses := svc.Status.LoadBalancer.Ingress
+	if len(ingresses) > 0 {
+		jsonPatch := fmt.Sprintf(`[{"op": "remove", "path": "/status/loadBalancer/ingress" }]`)
```
Is this a constant? Why are we using sprintf with no arguments?

balopat replied:
copy paste error, good catch

dlorenc commented on pkg/minikube/tunnel/loadbalancer_patcher.go:
```diff
@@ -0,0 +1,158 @@
+		logrus.Infof("Removed loadbalancer ingress from %s.", svc.Name)
+		return
+	} else {
+		return nil, nil
```
drop the else

dlorenc commented on pkg/minikube/tunnel/reporter.go:
```diff
@@ -0,0 +1,75 @@
+}
+
+func (r *SimpleReporter) Report(tunnelState *types.TunnelState) {
+	if reflect.DeepEqual(r.lastState, tunnelState) {
```
reflect is generally frowned upon outside of testing code. You should be able to use simple equality here on the struct values.

balopat replied:
fixed

balopat commented:
Couple of things while I was working on the comments: 
1.) I found a bug in the implementation: the daemon currently survives while I stop, delete and recreate a machine however doesn't pick up the new ip address. I either have to change the design to pick up the new ip address or to share fate with the machine (stop the daemon when minikube stops). 
2.) as a newbie golanger I overdesigned the abstractions for testability and created unnecessarily large amount of interface code for DI. As I'm reading more about this, there seems to be a lot of opportunity to simplify by passing `func` types around instead. 
A larger rework is coming soon.

balopat commented:
Removed the mocking framework and rolled my own stubs and fakes. 
I introduced the tunnel registry in order to be able to cleanup for orphaned tunnels. 
I still have some outstanding tasks: 
- [ ] refactor and validate windows version 
- [ ] refactor and validate linux version

tstromberg commented on pkg/minikube/tunnel/registry_test.go:
```diff
@@ -0,0 +1,275 @@
+		},
+		{
+			name: "Register + Remove + List",
+			test: func(t *testing.T, reg *persistentRegistry) {
```
Table driven tests shouldn't need custom per-test logic. Split these out into separate tests.

tstromberg replied:
Still this. table driven tests should never contain test-specific logic: http://go/gotestcomments#when-to-use-table-driven-tests

tstromberg commented on pkg/minikube/tunnel/registry_test.go:
```diff
@@ -0,0 +1,275 @@
+
+func createTestRegistry(t *testing.T) (reg *persistentRegistry, cleanup func()) {
+	f, err := ioutil.TempFile(os.TempDir(), "reg_")
+	f.Close()
```
Check error?

tstromberg commented on pkg/minikube/tunnel/process.go:
```diff
@@ -0,0 +1,53 @@
+	if runtime.GOOS == "windows" {
+		return err == nil, nil
+	}
+	//on unix systems further checking is required, as findProcess is noop
```
Also, runtime.GOOS checks feels sort of ugly: Go provides a mechanism for building different implementations for different platforms via build tags and file suffixes:

https://dave.cheney.net/2013/10/12/how-to-use-conditional-compilation-with-the-go-build-tool

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+	command := exec.Command("sudo", "route", "-n", "delete", route.DestCIDR.String())
+	if stdInAndOut, e := command.CombinedOutput(); e != nil {
+		return e
+	} else {
```
no need for else clause: you've already returned.

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+	}
+	command := exec.Command("sudo", "route", "-n", "delete", route.DestCIDR.String())
+	if stdInAndOut, e := command.CombinedOutput(); e != nil {
+		return e
```
You may want to return the output within the error message as well, because I suspect "e" doesn't have enough data to be debuggable.

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+	} else {
+		message := fmt.Sprintf("%s", stdInAndOut)
+		glog.V(4).Infof("%s", message)
+		re := regexp.MustCompile(fmt.Sprintf("^delete net ([^:]*)$"))
```
Remove unnecessary call to Sprintf().

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+		message := fmt.Sprintf("%s", stdInAndOut)
+		glog.V(4).Infof("%s", message)
+		re := regexp.MustCompile(fmt.Sprintf("^delete net ([^:]*)$"))
+		if !re.MatchString(message) {
```
Is this true? I haven't looked at the manpage yet, but it should be enough to check the error code. The output may be localized based on the users LANG settings, and is not meant to be an API.

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+		}
+		fields := strings.Fields(line)
+
+		if len(fields) > 2 {
```
Exit early if fields is less than 2.

tstromberg commented on pkg/minikube/tunnel/route_darwin.go:
```diff
@@ -0,0 +1,168 @@
+	return
+}
+
+func (router *osRouter) parseTable(table string) routingTable {
```
Should this communicate upwards if the table fails to be parsed at all?

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+	glog.Infof("About to run command: %s", command.Args)
+	stdInAndOut, e := command.CombinedOutput()
+	message := fmt.Sprintf("%s", stdInAndOut)
+	if len(message) > 0 {
```
Checking the error code should be sufficient here, unless you've seen otherwise in the API.

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+	if e != nil {
+		return e
+	}
+	if exists {
```
Should this check if the values are as intended?

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+/*
```
I was going to suggest golang.org/x/net/route - but it seems woefully incomplete for your usage :(

balopat replied:
yeah, I looked at syscall level integration as well, but it did not seemed simpler than calling out to netstat

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+	gatewayIP := route.Gateway.String()
+
+	glog.Infof("Cleaning up Route for CIDR %s to gateway %s\n", serviceCIDR, gatewayIP)
+	command := exec.Command("sudo", "ip", "route", "delete", serviceCIDR)
```
When parsing UNIX command output, set LC_ALL=C via command.Env. Otherwise you have to handle formatting for other languages.

balopat replied:
done

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+	stdInAndOut, e := command.CombinedOutput()
+	message := fmt.Sprintf("%s", stdInAndOut)
+	glog.Infof("%s", message)
+	if e != nil || len(message) > 0 {
```
Don't check the text output - it isn't a stable API. Why should we care if the ip implementation prints "Happy Birthday?". Just the error code.

balopat replied:
I guess that's true for Linux but not for MacOS - it was a copy paste, as I started on Mac. On Mac it always returns 0 and it prints an error when it fails :(

tstromberg commented on pkg/minikube/tunnel/route_test.go:
```diff
@@ -0,0 +1,139 @@
+func TestRoutingTable(t *testing.T) {
+	tcs := []struct {
+		name         string
+		routingTable routingTable
```
to avoid the stuttering, maybe just name the member 'table'?

balopat replied:
done

tstromberg commented on pkg/minikube/tunnel/route_linux.go:
```diff
@@ -0,0 +1,136 @@
+		err = fmt.Errorf("error running '%v': %s", command, e)
+		return
+	}
+	routeTableString := fmt.Sprintf("%s", stdInAndOut)
```
Sprintf is a slow way to convert bytes to strings. Just use string(), but only if you must. I'd argue that parseTable should probably take a slice of bytes.

balopat commented on cmd/minikube/cmd/tunnel.go:
```diff
@@ -0,0 +1,87 @@
+		RootCmd.PersistentPreRun(cmd, args)
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		flag.Lookup("logtostderr").Value.Set("true")
```
yeah, I was unsure about logging. If I leave this out, there is not much output. But using `fmt.Print*` felt "hacky" coming from the Java background - System.out.println is a no-no as that cannot be controlled by any form of loglevel setting.

balopat replied:
removed

balopat commented on cmd/minikube/cmd/tunnel.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func init() {
+	tunnelCmd.Flags().BoolVarP(&cleanup, "cleanup", "c", false, "call with cleanup=true to remove old tunnels")
```
I guess we can change it to auto-cleanup and try cleaning up every time, as cleanup does leave the useful tunnels (still running ones) in place. I was just operating under the assumption that `tunnel --cleanup` would be a separate command for handling the "unclean shutdown" cases - which should be rare.

balopat replied:
The cleanup command is supposed to only cleanup routes that are exact matches of minikube CIDR -> minikube IP and the process should not be running. The chances that a user has created an exact same route that she doesn't want deleted is minimal, so I guess we can safely include this at the start as well. 

However a standalone cleanup command is still useful to have without having to start a tunnel command - `minikube tunnel --cleanup` does just that.

balopat commented on cmd/minikube/cmd/update-context.go:
```diff
@@ -43,12 +43,13 @@ var updateContextCmd = &cobra.Command{
 			cmdUtil.MaybeReportErrorAndExit(err)
 		}
-		kstatus, err := kcfg.UpdateKubeconfigIP(ip, constants.KubeconfigPath, config.GetMachineName())
+		kstatus, err := kcfg.UpdateKubeconfigIP(ip, constants.KubeconfigPath, machineName)
```
ack

balopat commented on docs/networking.md:
```diff
@@ -10,3 +10,40 @@ To determine the NodePort for your service, you can use a `kubectl` command like
+INFO[0000] Creating k8s client...                       
+INFO[0000] Setting up router...                         
+INFO[0000] Setting up tunnel...                         
+INFO[0000] Started minikube tunnel. Tunnel requires root access. Please wait for the password prompt! 
```
true, it doesn. Also, this is actually old, I have removed this line since. I should remove it from the docs as well. Good catch!

balopat commented on pkg/minikube/cluster/cluster.go:
```diff
@@ -164,7 +164,7 @@ func GetHostDriverIP(api libmachine.API) (net.IP, error) {
 	}
 	ip := net.ParseIP(ipStr)
 	if ip == nil {
-		return nil, errors.Wrap(err, "Error parsing IP")
```
I mean, that's exactly what I was doing, just with `fmt.Errorf`- you commented on the old line?

balopat commented on pkg/minikube/tunnel/cluster_inspector.go:
```diff
@@ -0,0 +1,103 @@
+	"net"
+)
+
+type minikubeInspector struct {
```
I get what you're saying but `inspector` in itself would be easy to  misunderstand as tunnelInspector - and in fact it inspects the cluster...maybe`clusterInspector`?

balopat commented on pkg/minikube/constants/constants.go:
```diff
@@ -87,6 +87,10 @@ const DefaultStorageClassProvisioner = "standard"
 // Used to modify the cache field in the config file
 const Cache = "cache"
 
+func GetTunnelRegistryFile() string {
```
went with `TunnelRegistryPath` - the file itself is inconsistent in case of both Path/File and Get/Make/empty

balopat commented on hack/jenkins/unix_white_box_integration_tests.sh:
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
It is related, the whitebox tests were introduced here: https://github.com/kubernetes/minikube/pull/3015/files#diff-8dc4b2b9f0d2953af8122c32608048a3R31

balopat replied:
the rest, boilerplate and integration test related nits I did remove and will open a separate PR

balopat replied:
https://github.com/kubernetes/minikube/pull/3212

balopat commented on pkg/minikube/tunnel/process.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+func osGetPid() int {
+	return os.Getpid()
```
for testing purposes!

balopat commented on pkg/minikube/tunnel/process.go:
```diff
@@ -0,0 +1,53 @@
+		return err == nil, nil
+	}
+	//on unix systems further checking is required, as findProcess is noop
+	if err != nil {
```
FindProcess does give me a nice Process struct though. Is there no value in there?

tstromberg commented on pkg/minikube/tunnel/registry.go:
```diff
@@ -0,0 +1,185 @@
+			return e
+		}
+	}
+	defer f.Close()
```
If the content of this JSON file has any importance, check the error from f.Close().

tstromberg commented on pkg/minikube/tunnel/registry.go:
```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		return fmt.Errorf("error removing tunnel %s", err)
+	}
+	defer f.Close()
```
Check the error here too.

tstromberg commented on pkg/minikube/tunnel/registry_test.go:
```diff
@@ -0,0 +1,275 @@
+		},
+		{
+			name: "Register + Remove + List",
+			test: func(t *testing.T, reg *persistentRegistry) {
```
Name your errors 'err' when possible.

tstromberg commented on pkg/minikube/tunnel/tunnel.go:
```diff
@@ -0,0 +1,160 @@
+func (t *minikubeTunnel) updateTunnelStatus() *Status {
+	glog.V(3).Info("updating tunnel status...")
+	t.status.MinikubeState, _, t.status.MinikubeError = t.clusterInspector.getStateAndHost()
+	//TODO(balintp): clean this up to be more self contained
```
Clean this up, or remove the TODO. The grossness is self-evident.

tstromberg commented on pkg/minikube/tunnel/tunnel.go:
```diff
@@ -0,0 +1,160 @@
+	machineAPI libmachine.API,
+	configLoader config.Loader,
+	v1Core v1.CoreV1Interface, registry *persistentRegistry, router router) (*minikubeTunnel, error) {
+	clusterInspector := &clusterInspector{
```
name this "ci" so we don't have to guess if you mean the struct name or the local variable.

tstromberg commented on pkg/minikube/tunnel/tunnel.go:
```diff
@@ -0,0 +1,160 @@
+	return fmt.Errorf("there is already a running tunnel for this machine: %s", id)
+}
+
+func newTunnel(machineName string,
```
Keep function signatures as a single line when possible. When it isn't possible, there is your hint that you should take a config struct.

tstromberg commented on pkg/minikube/tunnel/tunnel.go:
```diff
@@ -0,0 +1,160 @@
+
+}
+
+type minikubeTunnel struct {
```
It's difficult to tell the difference between the `minikubeTunnel` struct and `tunnel` struct. Can you adjust one of the names to be more distinct?

`minikube.tunnel.minikubeTunnel` is a rather perverse naming stutter, but since the struct name is private I won't worry too much about this.

tstromberg commented on pkg/minikube/tunnel/tunnel.go:
```diff
@@ -0,0 +1,160 @@
+
+		exists, conflict, _, err := t.router.Inspect(t.status.TunnelID.Route)
+		if err != nil {
+			t.status.RouteError = fmt.Errorf("error checking for route state: %s", err)
```
Can this call return now? That would help to de-indent the non-error flow.

Same elsewhere in this function. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

tstromberg commented on pkg/minikube/tunnel/tunnel_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+func NewManager() *Manager {
+	return &Manager{
+		delay: 5 * time.Second,
```
Move this magic number to a const so that it's self-documenting.

tstromberg commented:
Approved with minor nitpicks. Thanks!

paultiplady commented:
Looking forward to giving this a spin, thanks for the hard work here!

omeid commented:
This looks very interesting and I am keen to give it a crack, any chance this could be cut into a release?
